src/bn254/tests/ec_pairing.rs

```diff
@@ -0,0 +1,381 @@
+
+        // Printing the number of constraints
+        let cs = owned_cs.into_assembly::<std::alloc::Global>();
+        cs.print_gate_stats();
+    }
+
+    /// Tests the torus final exponentiation step used in the pairing computation
+    /// which uses the torus compression.
+    ///
+    /// The test cases are loaded from the [`FINAL_EXP_TEST_CASES`] constant.
+    #[test]
+    #[ignore = "too-large circuit, should be run manually"]
+    fn test_final_exponentiation_torus() {
+        // Preparing the constraint system and parameters
+        let mut owned_cs = create_test_cs(1 << 25);
+        let cs = &mut owned_cs;
+
+        // Running tests from file
+        for (i, test) in FINAL_EXP_TEST_CASES.tests.iter().enumerate() {
+            // Expected:
+            let expected_f_final = test.expected.to_fq12(cs);
+
+            // Actual:
+            let mut f = test.scalar.to_fq12(cs);
+            let f_final = FinalExpEvaluation::evaluate_torus(cs, &mut f);
+            let f_final = f_final.get();
+
+            assert_equal_fq12(cs, &f_final, &expected_f_final);
+
+            println!(
+                "Final exponentiation with torus compression test {} has passed!",
+                i
+            );
+        }
+    }
+
+    #[test]
+    #[ignore = "used for debugging performance"]
+    fn debug_final_exponentiation_torus_performance() {
+        // Preparing the constraint system and parameters
+        let mut owned_cs = create_test_cs(1 << 21);
+        let cs = &mut owned_cs;
+
+        // Input:
+        let test_case = &FINAL_EXP_TEST_CASES.tests[0];
+        let mut f = test_case.scalar.to_fq12(cs);
+
+        // Performing the actual computation:
+        let _ = FinalExpEvaluation::evaluate_torus(cs, &mut f);
+
+        // Printing the number of constraints
+        let cs = owned_cs.into_assembly::<std::alloc::Global>();
+        cs.print_gate_stats();
+    }
+
+    /// Tests the EC pairing as a whole.
+    ///
+    /// The test cases are loaded from the [`PAIRING_TEST_CASES`] constant.
+    #[test]
+    #[ignore = "too-large circuit, should be run manually"]
+    fn test_ec_pairing() {
+        // Preparing the constraint system and parameters
+        let mut owned_cs = create_test_cs(1 << 21);
+        let cs = &mut owned_cs;
+
+        // Running tests from file
+        for (i, test) in PAIRING_TEST_CASES.tests.iter().enumerate() {
+            // Input:
+            let mut g1_point = test.g1_point.to_projective_point(cs);
+            let mut g2_point = test.g2_point.to_projective_point(cs);
+
+            // Expected:
+            let mut expected_pairing = test.pairing.to_fq12(cs);
+
+            // Actual:
+            let mut pairing = ec_pairing(cs, &mut g1_point, &mut g2_point);
+
+            // Asserting
+            assert_equal_fq12(cs, &mut pairing, &mut expected_pairing);
+
+            println!("EC pairing test {} has passed!", i);
+        }
+    }
+}
```

Comment:
might also be useful to include a test which ensures the bilinearity property holds with the created gadgets